File: backend/src/common/api-response.ts
export interface ApiEnvelope<T> {
  status: boolean;
  code: number;
  message: string;
  data: T;
}

export function ok<T>(data: T, message = "نجاح", code = 200): ApiEnvelope<T> {
  return {
    status: true,
    code,
    message,
    data,
  };
}

export function unwrapData<T extends object>(body: unknown): T {
  if (body && typeof body === "object" && "data" in body) {
    return (body as { data: T }).data;
  }
  return body as T;
}

File: lib/core/network/api_response.dart
class ApiResponse<T> {
  final bool status;
  final int? code;
  final String message;
  final T? data;
  final Map<String, dynamic> raw;

  const ApiResponse({
    required this.status,
    required this.message,
    required this.raw,
    this.code,
    this.data,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? value)? fromData,
  ) {
    return ApiResponse<T>(
      status: json['status'] == true,
      code: _asInt(json['code']),
      message: json['message']?.toString() ?? '',
      data: fromData == null ? json['data'] as T? : fromData(json['data']),
      raw: json,
    );
  }

  static int? _asInt(Object? value) {
    if (value is int) return value;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value);
    return null;
  }
}
